cdk/src/functions/workmail-organization.on-event-handler.ts

```diff
@@ -0,0 +1,138 @@
+export const PROP_PARAM_NAME = 'PropagationParamName';
+export const PROP_HOSTED_ZONE_ID = 'HostedZoneId';
+
+const workmail = new WorkMailClient({ region: 'eu-west-1' });
+const route53 = new Route53Client();
+
+export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  const domain = event.ResourceProperties[PROP_DOMAIN];
+  const hostedZoneId = event.ResourceProperties[PROP_HOSTED_ZONE_ID];
+
+  switch (event.RequestType) {
+    case 'Create':
+    case 'Update':
+      console.log(`${event.RequestType} Workmail organization. PhysicalResourceId: ${event.RequestId}`);
+
+      if (event.RequestType === 'Update') {
+        // Check if Workmail Organization exists
+        const orgList = await workmail.send(new ListOrganizationsCommand({}));
+
+        for (const org of orgList.OrganizationSummaries!) {
+          if (org.OrganizationId == event.PhysicalResourceId && org.DefaultMailDomain == domain) {
+            console.log('Workmail organization already exists, do nothing.');
+            return {
+              PhysicalResourceId: event.PhysicalResourceId,
+              Data: {
+                workmailOrgId: event.PhysicalResourceId,
+              },
+            };
+          }
+        }
+      }
+
+      console.log('Creating new workmail organization');
+      const command = new CreateOrganizationCommand({
+        Alias: uuidv4(),
+        Domains: [
+          {
+            DomainName: domain,
+            HostedZoneId: hostedZoneId,
+          },
+        ],
+        EnableInteroperability: false,
+      });
+
+      const response = await workmail.send(command);
+
+      return {
+        PhysicalResourceId: response.OrganizationId,
+        Data: {
+          workmailOrgId: response.OrganizationId,
+        },
+      };
+
+    case 'Delete':
+      console.log(`${event.RequestType} Workmail organization custom resource. PhysicalResourceId: ${event.PhysicalResourceId}`);
+
+      console.log(`Deleting Workmail / SES DNS records from hosted zone ${hostedZoneId}`);
+      await deleteDNSRecords(event.PhysicalResourceId!, hostedZoneId, domain);
+
+      console.log(`Deleting Workmail organization ${event.PhysicalResourceId}`);
+      await workmail.send(
+        new DeleteOrganizationCommand({
+          OrganizationId: event.PhysicalResourceId,
+          DeleteDirectory: true,
+          ForceDelete: false,
+        }),
+      );
+
+      return {
+        PhysicalResourceId: event.PhysicalResourceId,
+      };
+  }
+}
+
+async function deleteDNSRecords(workmailOrgId: string, hostedZoneId: string, domain: string): Promise<Object> {
+  // Get DNS records from Workmail
+  const workmail_domain = await workmail.send(
+    new GetMailDomainCommand({
+      OrganizationId: workmailOrgId,
+      DomainName: domain,
+    }),
+  );
+
+  // Parse the records for Route53 delete command input
+  const changeRecords: Change[] = [];
+
+  workmail_domain.Records!.forEach((record) => {
```

Comment:
where do you actually set the required DNS records for the create/update case?
You seem to be only deleting them. When I rolled this out in the pipeline account the DNS records were not created and mail got rejected.

Reply:
The DNS records were automatically created by workmail in the hosted zone when I deploy the stack in standalone-04. I will deploy it again on pipeline02 to see the error logs. If it indeed is the issue, I will create cdk resources for the DNS records.